### ISSUE TITLE
Support CAP negotiation in the Client

### ISSUE THREAD

**Issue by belak** [go-irc/irc#53]

**Comment by danopia** [comment]
What's the thinking on this as far as API goes?

**Comment by belak** [comment]
That's the main reason I haven't worked on this yet. There are a few complications, as there can be conflicting `CAP REQ`s and it's a different process if you're making the requests as the first thing or doing it in the middle of a session.

I'd love to hear any suggestions.

**Comment by belak** [comment]
The original idea was something like `c.CapReq(name string, required bool)` as it should be possible to have an optional CAP, but I'm sure there's a better solution than this.

**Comment by danopia** [comment]
I don't think you have to have special considerations for mid-session CAP as apps can already send the CAP REQ without any ceremony. The complication is that pre-session CAP (like SASL, I see there's a new ticket for that) has to happen _before_ the NICK/USER handshake, which the library has no way of disabling

**Comment by belak** [comment]
Right. Currently, the only way to do that would be by implementing it yourself using the Conn.

I'll see about adding support for `c.CapReq(name string, required bool) error`. This will queue up CAPs to request before sending the NICK/USER handshake and will error if conflicting CAPs are requested (and required) such as `multi-prefix` and `-multi-prefix`.

**Comment by belak** [comment]
Current plan (It's mostly implemented, I'm just ironing out a few oddities):

```
// CapReq represents the functions which will be added to the Client type
type CapReq interface {
    // CapRequest will add the given cap to a list of CAPs to request when the
    // client is run. Required will make the client error on startup if that cap is
    // not able to be enabled. This function will only have an effect if the client
    // has not been started yet.
    CapRequest(cap string, required bool)

    // CapEnabled can be called at any time and the current state of that CAP
    // will be returned.
    CapEnabled(cap string) bool

    // CapAvailable will return if this server supports the given CAP.
    CapAvailable(cap string) bool
}
```

**Comment by belak** [comment]
The main questions at this point:

* Should messages be sent to the handlers during CAP negotiation? 
* Do we need to respond to pings during CAP negotiation?
* Should the ping loop and send limiter be started before or after CAP negotiation?

**Comment by belak** [comment]
This should be resolved now that #58 has been merged in. I'm going to avoid tagging a new release until I get a chance to make sure it's reasonably stable, but this should be good to go.
